update freshness metric

emmadickson opened the pull request:
Add metric for the jenkins extracts. In support of adding more alerts to our Fulcrum pipeline. 

Missing 'fx' and 'criteo' extracts because those seem to go directly to redshift. 

Supports https://artsyproduct.atlassian.net/browse/PLATFORM-4326

joeyAghion commented on lib/fulcrum_extract_gravity_freshness.rb:
```diff
@@ -0,0 +1,65 @@
+    users: { bucket: 'artsy-data', prefix: 'reports/gravity_users' },
+    follow_artists: { bucket: 'artsy-data', prefix: 'reports/gravity_follow_artists' },
+    artwork_versions: { bucket: 'artsy-data', prefix: 'reports/gravity_artwork_versions' },
+    artworks: { bucket: 'artsy-data', prefix: 'reports/gravity_artworks' },
```
I think the prefixes are actually different, like `reports/gravity.artworks/`. Also, the trailing slash is important so that a path like `...artist_series` doesn't match others like `...artist_series_artworks`.

joeyAghion commented on lib/fulcrum_extract_gravity_freshness.rb:
```diff
@@ -0,0 +1,65 @@
+
+      age = Time.now - last_modified
+      warn "Recording freshness for extract #{key} as #{last_modified} with an age of #{age}"
+      statsd.gauge "postgres_extract_freshness.#{key}", age # seconds
```
Incorrect label here.

joeyAghion commented on lib/fulcrum_extract_non_gravity_freshness.rb:
```diff
@@ -0,0 +1,69 @@
+
+      age = Time.now - last_modified
+      warn "Recording freshness for extract #{key} as #{last_modified} with an age of #{age}"
+      statsd.gauge "gravity_extract_freshness.#{key}", age # seconds
```
And here.

joeyAghion reviewed (changes requested):
This appears to reopen the `DataFreshness` class, overwriting its constants and methods each time. I.e., I believe it won't work or will have unexpected results depending completely on which file was loaded last.

It also duplicates the S3 and statsd logic. Why? Is that in order to customize the metric namespace? It doesn't seem that important for attributes like "non-gravity" to be in the top-level namespace. These could instead be handled within the existing `DataFreshness` class and namespace by simply adding to the list of locations, like:

```
gravity_sales: { bucket: 'artsy-data', prefix: 'reports/gravity.sales/' },
```

Finally, the new work isn't triggered anywhere. Ultimately it will have to happen on a schedule within one of the [existing tasks](https://github.com/artsy/frequency/blob/main/Rakefile) or a new one. (If the new locations are simply added to the existing list as suggested above though, this will be automatic.)

emmadickson commented:
@joeyAghion I updated the pr to just add the s3 paths we want to track to the existing data freshness class. I also went back and fixed the confusing path names. I think this should be in better shape now.

joeyAghion commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,93 @@ class DataFreshness
-    user_price_preference: { bucket: 'artsy-cinder-production', prefix: 'builds/UserPricePreference/' }
+    user_price_preference: { bucket: 'artsy-cinder-production', prefix: 'builds/UserPricePreference/' },
+    fair_organizers: { bucket: 'artsy-data', prefix: 'reports/gravity.fair_organizers/' },
+    artist_series_artworks: { bucket: 'artsy-data', prefix: 'reports/gravity.partner_subscription_plans/' },
```
Incorrect key here.

joeyAghion commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,93 @@ class DataFreshness
+    exchange_admin_notes: { bucket: 'artsy-data', prefix: 'reports/exchange_admin_notes/' },
+    exchange_shipping_quote_requests: { bucket: 'artsy-data', prefix: 'reports/exchange_shipping_quote_requests/' },
+    exchange_quotes: { bucket: 'artsy-data', prefix: 'reports/exchange_shipping_quotes/' },
+    exchange_shipments: { bucket: 'artsy-data', prefix: 'reports/exchange_shipments/' },
```
These prefixes don't appear to match what's in S3.

joeyAghion commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,93 @@ class DataFreshness
+    marketo_form_submissions: { bucket: 'artsy-data', prefix: 'reports/marketo_form_submissions/' },
+    diffusion_lots_archive: { bucket: 'artsy-data', prefix: 'reports/diffusion.lots_archive/' },
+    diffusion_lots: { bucket: 'artsy-data', prefix: 'reports/diffusion.lots/' },
+    diffusion_source_lots_archive: { bucket: 'artsy-data', prefix: 'reports/diffusion.source_lots_archive/' },
```
Neither of the `diffusion.*_archive/` paths seem active. Perhaps they were left over from a migration?

emmadickson replied:
yup fixed this.

joeyAghion reviewed (commented):
Generally speaking, for the Gravity extracts, I think the keys should be namespaced with `gravity_` as with the other metrics, just for clarity and consistency. (Probably also alphabetized for easier scanning, once they're all grouped together in a namespace.)

I haven't verified that all of these are actively loaded into Redshift, but I commented about a few that I think are _not_. It might be worth checking this list against the load operations.

I'm not sure we should check 80+ S3 paths unless we suspect they might fail individually. How long does this take as currently written?

emmadickson commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,93 @@ class DataFreshness
+    convection_submissions: { bucket: 'artsy-data', prefix: 'reports/convection.submissions/' },
+    convection_users: { bucket: 'artsy-data', prefix: 'reports/convection.users/' },
+    convection_notes: { bucket: 'artsy-data', prefix: 'reports/convection.notes/' },
+    causality_high_bids: { bucket: 'artsy-data', prefix: 'reports/causality_high_bids/' },
```
I removed some paths that seem to not be updating to s3. I think they're still technically being called in the extract though so I want to go back and look more closely at the fulcrum extract.

joeyAghion reviewed (commented):
Definitely think it needs alphabetization now.

emmadickson commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,88 @@ class DataFreshness
+    causality_events: { bucket: 'artsy-data', prefix: 'reports/causality.events/' },
+    causality_lot_states: { bucket: 'artsy-data', prefix: 'reports/causality.lot_states/' },
+    positron_articles: { bucket: 'artsy-data', prefix: 'reports/positron.articles/' },
+    exchange_orders: { bucket: 'artsy-data', prefix: 'reports/exchange_production._orders/' },
```
fixed

joeyAghion commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,86 @@ class DataFreshness
+    exchange_shipping_quotes: { bucket: 'artsy-data', prefix: 'reports/exchange_production.shipping_quotes/' },
+    exchange_state_histories: { bucket: 'artsy-data', prefix: 'reports/exchange_production.state_histories/' },
+    exchange_transactions: { bucket: 'artsy-data', prefix: 'reports/exchange_production.transactions/' },
+    impulse_assessments: { bucket: 'artsy-data', prefix: 'reports/impulse_assessments/' },
```
Should be `impulse.assessments` I think.

joeyAghion commented on lib/data_freshness.rb:
```diff
@@ -28,7 +28,86 @@ class DataFreshness
+    impulse_conversation_items: { bucket: 'artsy-data', prefix: 'reports/impulse.conversation_items/' },
+    impulse_conversations: { bucket: 'artsy-data', prefix: 'reports/impulse.conversations/' },
+    impulse_messages: { bucket: 'artsy-data', prefix: 'reports/impulse.messages/' },
+    gravity_account_activities: { bucket: 'artsy-data', prefix: 'reports/gravity.account_activies/' },
```
Typo still.

emmadickson replied:
Fixed
